vortex-array/src/compute/binary_numeric.rs

```diff
@@ -0,0 +1,249 @@
+        BinaryNumericFn::binary_numeric(encoding, array_ref, rhs, op)
+    }
+}
+
+/// Point-wise add two numeric arrays.
+pub fn add(lhs: impl AsRef<ArrayData>, rhs: impl AsRef<ArrayData>) -> VortexResult<ArrayData> {
+    binary_numeric(lhs.as_ref(), rhs.as_ref(), NumericOperator::Add)
+}
+
+/// Point-wise add a scalar value to this array on the right-hand-side.
+pub fn add_scalar(lhs: impl AsRef<ArrayData>, rhs: Scalar) -> VortexResult<ArrayData> {
+    let lhs = lhs.as_ref();
+    binary_numeric(
+        lhs,
+        &ConstantArray::new(rhs, lhs.len()).into_array(),
+        NumericOperator::Add,
+    )
+}
+
+/// Point-wise subtract two numeric arrays.
+pub fn sub(lhs: impl AsRef<ArrayData>, rhs: impl AsRef<ArrayData>) -> VortexResult<ArrayData> {
+    binary_numeric(lhs.as_ref(), rhs.as_ref(), NumericOperator::Sub)
+}
+
+/// Point-wise subtract a scalar value from this array on the right-hand-side.
+pub fn sub_scalar(lhs: impl AsRef<ArrayData>, rhs: Scalar) -> VortexResult<ArrayData> {
+    let lhs = lhs.as_ref();
+    binary_numeric(
+        lhs,
+        &ConstantArray::new(rhs, lhs.len()).into_array(),
+        NumericOperator::Sub,
+    )
+}
+
+/// Point-wise multiply two numeric arrays.
+pub fn mul(lhs: impl AsRef<ArrayData>, rhs: impl AsRef<ArrayData>) -> VortexResult<ArrayData> {
+    binary_numeric(lhs.as_ref(), rhs.as_ref(), NumericOperator::Mul)
+}
+
+/// Point-wise multiply a scalar value into this array on the right-hand-side.
+pub fn mul_scalar(lhs: impl AsRef<ArrayData>, rhs: Scalar) -> VortexResult<ArrayData> {
+    let lhs = lhs.as_ref();
+    binary_numeric(
+        lhs,
+        &ConstantArray::new(rhs, lhs.len()).into_array(),
+        NumericOperator::Mul,
+    )
+}
+
+/// Point-wise divide two numeric arrays.
+pub fn div(lhs: impl AsRef<ArrayData>, rhs: impl AsRef<ArrayData>) -> VortexResult<ArrayData> {
+    binary_numeric(lhs.as_ref(), rhs.as_ref(), NumericOperator::Div)
+}
+
+/// Point-wise divide a scalar value into this array on the right-hand-side.
+pub fn div_scalar(lhs: impl AsRef<ArrayData>, rhs: Scalar) -> VortexResult<ArrayData> {
+    let lhs = lhs.as_ref();
+    binary_numeric(
+        lhs,
+        &ConstantArray::new(rhs, lhs.len()).into_array(),
+        NumericOperator::Mul,
+    )
+}
+
+pub fn binary_numeric(
+    lhs: &ArrayData,
+    rhs: &ArrayData,
+    op: NumericOperator,
+) -> VortexResult<ArrayData> {
+    if lhs.len() != rhs.len() {
+        vortex_bail!("Numeric operations aren't supported on arrays of different lengths")
+    }
+    if !matches!(lhs.dtype(), DType::Primitive(_, _))
+        || !matches!(rhs.dtype(), DType::Primitive(_, _))
+        || lhs.dtype() != rhs.dtype()
+    {
+        vortex_bail!(
+            "Numeric operations are only supported on two arrays sharing the same primitive-type: {} {}",
+            lhs.dtype(),
+            rhs.dtype()
+        )
+    }
+
+    // Check if LHS supports the operation directly.
```

Comment:
You should invert LHS and RHS for commutative operators (add, mul) and check if RHS supports LHS.